teuthology/orchestra/remote.py

```diff
@@ -199,14 +199,10 @@ def run(self, **kwargs):
 
         TODO refactor to move run.run here?
         """
-        if self.ssh is None:
-            self.reconnect(timeout=5)
-        if self.ssh:
-            # There is a chance that the ssh(paramiko) instance
-            # is populated but has no transport yet.
-            if self.ssh.get_transport():
-                if not self.ssh.get_transport().is_active():
-                    self.reconnect(timeout=5)
+        if not self.ssh or \
+           not self.ssh.get_transport() or \
+           not self.ssh.get_transport().is_active():
+            self.reconnect()
```

Comment:
I agree with the patch, it looks like possibility of missing transport wasn't counted before.

I don't quite agree to remove ```timeout=5``` . It looks like with some usual ```run``` it could happen that connection is stuck so it would give 5 sec before reconection.

For node reboot do you guys use reconnect before next run? As it is here"
```
            remote.run(args=['sudo', 'shutdown', '-r', 'now'], wait=False)
            time.sleep(40)
            log.info("Reconnecting after reboot")
            remote.reconnect(timeout=300)
            remote.run(args=['uname', '-a'])
```

Reply:
yeah, looks like remote_exec that is used by DeepSea task, doesn't do reconnect.
Probably
```
            except ConnectionLostError:
                already_rebooted_at_least_once = True
```

should have remote.reconnect with proper timeout in case it was supposed to reboot.

Reply:
> I agree with the patch, it looks like possibility of missing transport wasn't counted before.
> 
> I don't quite agree to remove `timeout=5` . It looks like with some usual `run` it could happen that connection is stuck so it would give 5 sec before reconection.

Don't agree, we should not use any timeout=5 because, the _reconnect_ method has  a bug, it is not related to this particular issue, so using timeout does not give us nothing.
For more  details, the see code of try loop. If you use any timeout < 30 the reconnect will always give it a reconnect try only once. From other side, socket.timeout is 60 seconds in the code and even if the 'connect' operation stuck giving timeout will give nothing, because it will not interrupt socket operation after 5 seconds. Since that, we should not submit code which is not working.

> 
> For node reboot do you guys use reconnect before next run? As it is here"
> 
> ```
>             remote.run(args=['sudo', 'shutdown', '-r', 'now'], wait=False)
>             time.sleep(40)
>             log.info("Reconnecting after reboot")
>             remote.reconnect(timeout=300)
>             remote.run(args=['uname', '-a'])
> ```

Reply:
> yeah, looks like remote_exec that is used by DeepSea task, doesn't do reconnect.
> Probably
> 
> ```
>             except ConnectionLostError:
>                 already_rebooted_at_least_once = True
> ```
> 
> should have remote.reconnect with proper timeout in case it was supposed to reboot.

 ```with safe_while(sleep=60, tries=tries, action="wait for reconnect") as proceed:```
ah, it looks like another re implementation of ```reconnect```